data/pools/transactionPool.go

```diff
@@ -74,18 +79,29 @@ type TransactionPool struct {
 	assemblyRound   basics.Round
 	assemblyResults poolAsmResults
 
-	// pendingMu protects pendingTxGroups and pendingTxids
-	pendingMu       deadlock.RWMutex
-	pendingTxGroups [][]transactions.SignedTxn
-	pendingTxids    map[transactions.Txid]transactions.SignedTxn
+	// pendingMu protects pendingTxGroups, pendingTxids, pendingCounter and pendingLatestLocal
+	pendingMu deadlock.RWMutex
+	// pendingTxGroups is a slice of the pending transaction groups.
+	pendingTxGroups []transactions.SignedTxGroup
+	// pendingTxids is a map of the pending *transaction ids* included in the pendingTxGroups array.
+	pendingTxids map[transactions.Txid]transactions.SignedTxn
+	// pendingCounter is a monotomic counter, indicating the next pending transaction group counter value.
+	pendingCounter uint64
+	// pendingLatestLocal is the value of the last transaction group counter which is associated with a transaction that was
+	// locally originated ( i.e. posted to this node via the REST API )
+	pendingLatestLocal uint64
 
 	// Calls to remember() add transactions to rememberedTxGroups and
 	// rememberedTxids.  Calling rememberCommit() adds them to the
 	// pendingTxGroups and pendingTxids.  This allows us to batch the
 	// changes in OnNewBlock() without preventing a concurrent call
-	// to PendingTxGroups() or Verified().
-	rememberedTxGroups [][]transactions.SignedTxn
+	// to PendingTxGroups().
+	rememberedTxGroups []transactions.SignedTxGroup
 	rememberedTxids    map[transactions.Txid]transactions.SignedTxn
+	// rememberedLatestLocal is the value of the last transaction group counter which is associated with a transaction that was
+	// locally originated ( i.e. posted to this node via the REST API ). This variable is used when OnNewBlock is called and
+	// we filter out the pending transaction through the evaluator.
+	rememberedLatestLocal uint64
```

Comment:
Seems like this is not needed. We can just update `pendingLatestLocal` in `rememberCommit()` instead. That would be simpler.

Reply:
That's already what we do. The reason we need the two is because when a new round arrives, we re-add all the entries from the `pending` back into the `remembered`. We need to have two distinguished counters so that we can query the transaction pool for the list of transactions *while* running the block evaluator.

Reply:
I'm not getting it. We read from `rememberedLatestLocal` only at line 248. Can we loop through `pendingTxGroups` there instead?

Reply:
sure we can - but we're going to make quite a long loop : we can have ( at worst case scenario ) a full transaction pool. That's quite many transactions.

Reply:
think of the `pendingTxGroups`/`pendingTxids`/`pendingLatestLocal` and their `remembered` counterparts as a "live" vs. "staging" buffers. We want to keep the "live" buffer available as long as we can.

Reply:
Ok, I see.